### ISSUE TITLE
Obtain Realm instance from RealmResults (and maybe managed RealmList)

### ISSUE THREAD

**Issue by Zhuinden** [realm/realm-java#5852]
#### Goal

> What do you want to achieve?

Obtain Realm from RealmResults in a RealmChangeListener.

#### Expected Results

I should be able to obtain it, without having to have the Realm instance nearby and separately.

#### Actual Results

I think only `RealmObject.getRealm()` exists, but not for RealmResults.

#### Code Sample

```java
private RealmChangeListener<RealmResults<Member>> objectTracker = (results) -> {
    if(member != null && member.isValid()) {
        return;
    }
    if(realm.isInTransaction()) { // <-- i can't get this from results
        handler.post(() -> {
            setupMember(results);
        });
    } else {
        setupMember(results);
    }
};
```

#### Version of Realm and tooling
Realm version(s): 5.0.1

Realm sync feature enabled: no

**Comment by cmelchior** [comment]
We have internally  talked about adding this, and back then we decided against it since it is kinda complicated since calling `getRealm()` on a `RealmResults<DynamicRealmObject>` doesn't make a lot of sense and having both `getRealm()` and `getDynamicRealm()` would be extremely annoying.

That said, that is then exactly what we did for `DynamicRealmObject` which has both `getRealm()` (which will throw) since it comes from `RealmObject` and `DynamicRealmObject.getDynamicRealm()`

It is in situations like this I wished we actually had completely separate types for `DynamicRealmResults` and normal `RealmResults`, but that would also blow up our API surface. Tradeoffs 🙄 

With that little bit of historical context and given what the API is today, it wouldn't be totally far fetched for us to add this, especially since the dynamic API never had the same amount of features as the typed one. The method should just throw if the Realm was backed by `DynamicRealm`. I think I could live with that, especially since it would be nice if you didn't have to parse a Realm instance around everywhere, but could just grap it from the object/result/list you had.
